docs/powerkit.1
.\" Automatically generated by Pandoc 3.1.8
.\"
.TH "POWERKIT" "1" "February 2024" "Version 2.0.0" "PowerKit Documentation"
.SH NAME
\f[I]powerkit\f[R] - Desktop independent power manager for Linux
.SH SYNOPSIS
powerkit \f[I]\f[CI][--config]\f[I]\f[R]
\f[I]\f[CI][--set-brightness-up]\f[I]\f[R]
\f[I]\f[CI][--set-brightness-down]\f[I]\f[R]
\f[I]\f[CI][--sleep]\f[I]\f[R] \f[I]\f[CI][--hibernate]\f[I]\f[R]
\f[I]\f[CI][--lock]\f[I]\f[R]
.SH DESCRIPTION
Desktop independent power manager for use with alternative X11 desktop
environments and window managers on Linux.
.IP \[bu] 2
Implements \f[I]\f[CI]org.freedesktop.ScreenSaver\f[I]\f[R] service
.IP \[bu] 2
Implements \f[I]\f[CI]org.freedesktop.PowerManagement.Inhibit\f[I]\f[R]
service
.IP \[bu] 2
Automatic actions on lid, idle and low battery
.RS 2
.IP \[bu] 2
Sleep
.IP \[bu] 2
Hibernate
.IP \[bu] 2
HybridSleep
.IP \[bu] 2
Suspend then Hibernate
.IP \[bu] 2
Shutdown
.RE
.IP \[bu] 2
Screen saver support
.IP \[bu] 2
Screen lid support
.IP \[bu] 2
Screen locking support
.IP \[bu] 2
Screen backlight support
.IP \[bu] 2
Notification support (can use
\f[I]\f[CI]org.freedesktop.Notifications\f[I]\f[R] if available)
.SH USAGE
powerkit should be started during the X11 user session.
Consult the documentation for your desktop environment or window manager
for launching startup applications.
.IP \[bu] 2
In \f[I]Fluxbox\f[R] add \f[I]\f[CI]powerkit &\f[I]\f[R] to the
\f[I]\f[CI]\[ti]/.fluxbox/startup\f[I]\f[R] file
.IP \[bu] 2
In \f[I]Openbox\f[R] add \f[I]\f[CI]powerkit &\f[I]\f[R] to the
\f[I]\f[CI]\[ti]/.config/openbox/autostart\f[I]\f[R] file.
.PP
\f[B]Do use powerkit if your desktop environment or window manager
already has a power manager or screen saver service running.\f[R]
.SS CONFIGURATION
Settings are available directly from the system tray icon or run
\f[I]\f[CI]powerkit --config\f[I]\f[R].
You should also be able to lauch the powerkit settings from your desktop
application menu (if available).
.SS SCREEN SAVER
powerkit implements a basic screen saver to handle screen blanking,
poweroff and locking feature.
.PP
Locking feature depends on \f[CR]xsecurelock\f[R].
.PP
You can override the lock command with
\f[I]\f[CI]screensaver_lock_cmd=<command>\f[I]\f[R] in
\f[I]\f[CI]\[ti]/.config/powerkit/powerkit.conf\f[I]\f[R].
Note that the command must not contain spaces.
.SS BACKLIGHT
The current display brightness (on laptops and supported displays) can
be adjusted with the mouse wheel on the system tray icon or through the
system tray menu.
.PP
powerkit also supports the following commands that can be used for
global shortcuts, scripts etc:
.TP
\f[I]\f[CI]powerkit --set-brightness-up\f[I]\f[R]
Set default display brightness up.
.TP
\f[I]\f[CI]powerkit --set-brightness-down\f[I]\f[R]
Set default display brightness down.
.PP
On Fluxbox you can add the commands to the
\f[I]\f[CI]\[ti]/.fluxbox/keys\f[I]\f[R] file:
.IP
.EX
XF86MonBrightnessUp :Exec powerkit --set-brightness-up
XF86MonBrightnessDown :Exec powerkit --set-brightness-down
.EE
.SS HIBERNATE
If hibernate works depends on your system, a swap partition (or file) is
needed by the kernel to support hibernate.
.PP
\f[B]\f[BI]Consult your system documentation regarding
hibernation\f[B]\f[R].
.SS ICONS
powerkit will use the existing icon theme from the running desktop
environment or window manager.
.PP
You can override the icon theme in the
\f[I]\f[CI]\[ti]/.config/powerkit/powerkit.conf\f[I]\f[R] file, use
\f[I]\f[CI]icon_theme=<theme_name>\f[I]\f[R].
.SH FAQ
.SS How does an application inhibit the screen saver?
The preferred way to inhibit the screen saver from an application is to
use the \f[I]org.freedesktop.ScreenSaver\f[R] specification.
Any application that uses \f[I]org.freedesktop.ScreenSaver\f[R] will
work with powerkit.
.PP
Popular applications that uses this feature is Mozilla Firefox, Google
Chrome, VideoLAN VLC and many more.
.SS How does an application inhibit suspend actions?
The preferred way to inhibit suspend actions from an application is to
use the \f[I]org.freedesktop.PowerManagement\f[R] specification.
Any application that uses \f[I]org.freedesktop.PowerManagement\f[R] will
work with powerkit.
.PP
Common use cases are audio playback, downloading, rendering and similar.
.SS Google Chrome/Chromium does not inhibit the screen saver or power manager!?
Chrome does not use \f[I]org.freedesktop.ScreenSaver\f[R] or
\f[I]org.freedesktop.PowerManagement\f[R] until it detects a supported
desktop environment.
Add the following to \f[I]\f[CI]\[ti]/.bashrc\f[I]\f[R] or the
\f[I]\f[CI]google-chrome\f[I]\f[R] launcher if you don\[cq]t run a
supported desktop environment:
.IP
.EX
export DESKTOP_SESSION=xfce
export XDG_CURRENT_DESKTOP=xfce
.EE
.SS Mozilla Firefox does not inhibit the power manager during audio playback and/or downloading!?
This is an issue with Firefox (missing feature).
Use a different browser or open a request on the Firefox issue tracker.
.PP
Firefox should inhibit the power manager during audio playback
(regardless of video) and during download (active queue).
Currently Firefox only inhibit the screen saver during video playback.
Chrome/Chromium does this correctly.
.SH REQUIREMENTS
powerkit requires the following dependencies:
.IP \[bu] 2
\f[I]X11 (https://www.x.org)\f[R]
.IP \[bu] 2
\f[I]libXss (https://www.x.org/archive//X11R7.7/doc/man/man3/Xss.3.xhtml)\f[R]
.IP \[bu] 2
\f[I]libXrandr (https://www.x.org/wiki/libraries/libxrandr/)\f[R]
.IP \[bu] 2
\f[I]Qt (https://qt.io)\f[R] 5.15 \f[I](Core/DBus/Gui/Widgets)\f[R]
.IP \[bu] 2
\f[I]logind (https://www.freedesktop.org/wiki/Software/systemd/logind/)\f[R]
\f[I](or compatible service)\f[R]
.IP \[bu] 2
\f[I]UPower (https://upower.freedesktop.org/)\f[R] \f[I](or compatible
service)\f[R]
.IP \[bu] 2
\f[I]xsecurelock (https://github.com/google/xsecurelock)\f[R]
.SH BUILD
First make sure you have the required dependencies installed, then
review the most common build options:
.IP \[bu] 2
\f[I]\f[CI]CMAKE_INSTALL_PREFIX=</usr/local>\f[I]\f[R] - Install target.
\f[I]\f[CI]/usr\f[I]\f[R] recommended.
.IP \[bu] 2
\f[I]\f[CI]CMAKE_BUILD_TYPE=<Release/Debug>\f[I]\f[R] - Build type.
\f[I]\f[CI]Release\f[I]\f[R] recommended
.PP
Now configure powerkit with CMake and build:
.IP
.EX
mkdir build && cd build
cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=Release ..
make -j4
.EE
.PP
All you need is the \f[I]\f[CI]powerkit\f[I]\f[R] binary located in the
build directory, you can run it from any location.
.SS Install
Use regular \f[I]\f[CI]make install\f[I]\f[R] with optional
\f[I]\f[CI]DESTDIR\f[I]\f[R]:
.IP
.EX
make DESTDIR=<package_directory> install
.EE
.PP
or make a native package:
.IP
.EX
cpack -G DEB
.EE
.IP
.EX
cpack -G RPM
.EE
.SH CHANGELOG
.SS 2.0.0 (TBA)
.IP \[bu] 2
Recommended locker is \f[CR]xsecurelock\f[R]
.IP \[bu] 2
Added support for \[lq]modern\[rq] logind
.IP \[bu] 2
Removed support for ConsoleKit
.IP \[bu] 2
Removed support for XScreenSaver
.IP \[bu] 2
Added basic screen saver
.IP \[bu] 2
Easier to use (minimal/no setup)
.IP \[bu] 2
New UI
.IP \[bu] 2
Major code changes
.SH OPTIONS
.TP
\f[I]\f[CI]--config\f[I]\f[R]
Launch configuration.
.TP
\f[I]\f[CI]--set-brightness-up\f[I]\f[R]
Set default display brightness up.
.TP
\f[I]\f[CI]--set-brightness-down\f[I]\f[R]
Set default display brightness down.
.TP
\f[I]\f[CI]--sleep\f[I]\f[R]
Suspend computer now.
Can be combined with \f[I]\f[CI]--hibernate\f[I]\f[R] for suspend then
hibernate after X amount of time.
.TP
\f[I]\f[CI]--hibernate\f[I]\f[R]
Hibernate computer now.
Can be combined with \f[I]\f[CI]--sleep\f[I]\f[R] for suspend then
hibernate after X amount of time.
.TP
\f[I]\f[CI]--lock\f[I]\f[R]
Lock screen.
.SH FILES
.TP
\f[I]\f[CI]\[ti]/.config/powerkit/powerkit.conf\f[I]\f[R]
Per user configuration file.
.SH SEE ALSO
\f[B]\f[CB]xsecurelock\f[B]\f[R](1), \f[B]\f[CB]UPower\f[B]\f[R](7),
\f[B]\f[CB]systemd-logind\f[B]\f[R](8)
.SH BUGS
See \f[B]https://github.com/rodlie/powerkit/issues\f[R].
.SH COPYRIGHT
.IP
.EX
Copyright (c) Ole-André Rodlie <https://github.com/rodlie>
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of the copyright holder nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \[dq]AS IS\[dq]
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.EE
.SH AUTHORS
Ole-André Rodlie.


ChangeLog
## 1.0.0rc2 - 20181103
 * Added more configuration options
 * Added option to bundle failsafe icons

## 1.0.0rc1 - 20181025
 * first public release


src/powerkit.cpp
/*
# PowerKit <https://github.com/rodlie/powerkit>
# Copyright (c) Ole-André Rodlie <https://github.com/rodlie> All rights reserved.
#
# Available under the 3-clause BSD license
# See the LICENSE file for full details
*/

#include <QApplication>

#include "powerkit_app.h"
#include "powerkit_dialog.h"
#include "powerkit_common.h"
#include "powerkit_backlight.h"
#include "powerkit_client.h"

#define CMD_OPT_CONFIG "--config"
#define CMD_OPT_BRIGHTNESS_UP "--set-brightness-up"
#define CMD_OPT_BRIGHTNESS_DOWN "--set-brightness-down"
#define CMD_OPT_SLEEP "--sleep"
#define CMD_OPT_HIBERNATE "--hibernate"
#define CMD_OPT_LOCK "--lock"

int main(int argc, char *argv[])
{
    QApplication a(argc, argv);
    QCoreApplication::setApplicationName("freedesktop");
    QCoreApplication::setOrganizationDomain("org");
    QCoreApplication::setApplicationVersion(QString::fromUtf8(POWERKIT_VERSION));

    const auto args = QApplication::arguments();
    bool openConfig = args.contains(CMD_OPT_CONFIG);
    bool setBrightness = (args.contains(CMD_OPT_BRIGHTNESS_UP) ||
                          args.contains(CMD_OPT_BRIGHTNESS_DOWN));
    bool setSleep = args.contains(CMD_OPT_SLEEP);
    bool setHibernate = args.contains(CMD_OPT_HIBERNATE);
    bool setLock = args.contains(CMD_OPT_LOCK);

    if (openConfig) {
        if (!QDBusConnection::sessionBus().registerService(POWERKIT_CONFIG)) {
            qWarning() << QObject::tr("A powerkit config instance is already running");
            return 1;
        }
        PowerKit::Dialog dialog;
        dialog.show();
        return a.exec();
    } else if (setBrightness) {
        int val = PowerKit::Backlight::getCurrentBrightness();
        if (args.contains(CMD_OPT_BRIGHTNESS_UP)) { val += POWERKIT_BACKLIGHT_STEP; }
        else if (args.contains(CMD_OPT_BRIGHTNESS_DOWN)) { val -= POWERKIT_BACKLIGHT_STEP; }
        return PowerKit::Backlight::setBrightness(val);
    } else if (setSleep || setHibernate || setLock) {
        QDBusInterface manager(POWERKIT_SERVICE,
                               POWERKIT_PATH,
                               POWERKIT_MANAGER,
                               QDBusConnection::sessionBus());
        if (!manager.isValid()) {
            qWarning() << QObject::tr("Unable to connect to the powerkit service");
            return 1;
        }
        if (setSleep && setHibernate) {
            return PowerKit::Client::suspendThenHibernate(&manager);
        } else if (setSleep) {
            return PowerKit::Client::suspend(&manager);
        } else if (setHibernate) {
            return PowerKit::Client::hibernate(&manager);
        } else if (setLock) {
            return PowerKit::Client::lockScreen(&manager);
        }
        return 1;
    }

    QDBusInterface session(POWERKIT_SERVICE,
                           POWERKIT_PATH,
                           POWERKIT_SERVICE,
                           QDBusConnection::sessionBus());
    if (session.isValid()) {
        qWarning() << QObject::tr("A powerkit session is already running");
        return 1;
    }

    PowerKit::App powerkit(a.parent());
    return a.exec();
}
